File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.kt
class Solution {
    fun solution(p: String): String {
        return convert(p)
    }

    private fun convert(p: String): String {
        if (p.isEmpty()) return ""

        // 1. 균형잡힌 문자열 u, v 분리
        var balance = 0
        var idx = 0
        for (i in p.indices) {
            if (p[i] == '(') balance++
            else balance--

            if (balance == 0) {
                idx = i
                break
            }
        }

        val u = p.substring(0, idx + 1)
        val v = p.substring(idx + 1)

        // 2. u가 올바른 괄호 문자열이면 u + convert(v)
        if (isCorrect(u)) {
            return u + convert(v)
        }

        // 3. u가 올바르지 않으면 규칙에 따라 변환
        val sb = StringBuilder()
        sb.append("(")
        sb.append(convert(v))
        sb.append(")")

        // u의 앞뒤 제거 후 뒤집기
        for (i in 1 until u.length - 1) {
            sb.append(if (u[i] == '(') ')' else '(')
        }

        return sb.toString()
    }

    // 올바른 괄호 문자열인지 검사
    private fun isCorrect(s: String): Boolean {
        var count = 0
        for (c in s) {
            if (c == '(') count++
            else count--

            if (count < 0) return false
        }
        return count == 0
    }
}

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.js
function solution(p) {
    return convert(p);
}

function convert(p) {
    if (p === "") return "";

    // 1. u, v 분리 (u는 균형잡힌 괄호 문자열)
    let balance = 0;
    let idx = 0;

    for (let i = 0; i < p.length; i++) {
        if (p[i] === "(") balance++;
        else balance--;

        if (balance === 0) {
            idx = i;
            break;
        }
    }

    const u = p.slice(0, idx + 1);
    const v = p.slice(idx + 1);

    // 2. u가 올바른지 확인
    if (isCorrect(u)) {
        return u + convert(v);
    }

    // 3. u가 올바르지 않은 경우 규칙대로 새 문자열 생성
    let result = "(";
    result += convert(v);
    result += ")";

    // 4. u의 첫/마지막을 제거하고 나머지 뒤집기
    let flipped = "";
    for (let i = 1; i < u.length - 1; i++) {
        flipped += u[i] === "(" ? ")" : "(";
    }

    return result + flipped;
}

// 올바른 괄호 문자열인지 검사
function isCorrect(s) {
    let count = 0;
    for (let c of s) {
        if (c === "(") count++;
        else count--;

        if (count < 0) return false;
    }
    return count === 0;
}
